`BallerinaErrorHandler` reaching fail-safe for incomplete code sample

**Description:**

Below sample is reaching fail-safe  in `BallerinaErrorHandler`
```bal
var x = from var a in m = {
```
<details>
  <summary>Stack trace.(Click to expand)</summary>

```
Exception in thread "main" java.lang.AssertionError: fail safe reached
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.recover(AbstractParserErrorHandler.java:115)
	at io.ballerina.compiler.internal.parser.AbstractParser.recover(AbstractParser.java:111)
	at io.ballerina.compiler.internal.parser.BallerinaParser.recover(BallerinaParser.java:19067)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseIntermediateClause(BallerinaParser.java:11932)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseQueryExprRhs(BallerinaParser.java:11809)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTableConstructorOrQueryInternal(BallerinaParser.java:11726)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTableConstructorOrQuery(BallerinaParser.java:11716)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTerminalExpression(BallerinaParser.java:5134)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTerminalExpression(BallerinaParser.java:5065)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTerminalExpression(BallerinaParser.java:5059)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseExpression(BallerinaParser.java:5037)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseExpression(BallerinaParser.java:5032)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseExpression(BallerinaParser.java:5027)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseExpression(BallerinaParser.java:4958)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseVarDeclRhs(BallerinaParser.java:4628)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseVariableDecl(BallerinaParser.java:4567)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseModuleVarDecl(BallerinaParser.java:1084)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:1038)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:984)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:499)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:431)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseCompUnit(BallerinaParser.java:335)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parse(BallerinaParser.java:100)
	at io.ballerinalang.compiler.parser.test.TestParser.testParser(TestParser.java:46)
	at io.ballerinalang.compiler.parser.test.TestParser.main(TestParser.java:40)
```
</details>

--- a/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java
+++ b/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java
@@ -2744,7 +2744,7 @@ public class BallerinaParser extends AbstractParser {
                     return parseParameterizedTypeDescriptor(consume());
                 }
                 
-                if (isSingletonTypeDescStart(nextToken.kind)) {
+                if (isSingletonTypeDescStart(nextToken.kind, getNextNextToken())) {
                     reportInvalidQualifierList(qualifiers);
                     return parseSingletonTypeDesc();
                 }
@@ -9857,7 +9857,11 @@ public class BallerinaParser extends AbstractParser {
     }
 
     private boolean isTypeStartingToken(SyntaxKind nodeKind) {
-        switch (nodeKind) {
+        return isTypeStartingToken(nodeKind, getNextNextToken());
+    }
+    
+    static boolean isTypeStartingToken(SyntaxKind nextTokenKind, STToken nextNextToken) {
+        switch (nextTokenKind) {
             case IDENTIFIER_TOKEN:
             case SERVICE_KEYWORD:
             case RECORD_KEYWORD:
@@ -9876,14 +9880,14 @@ public class BallerinaParser extends AbstractParser {
             case TRANSACTION_KEYWORD:
                 return true;
             default:
-                if (isParameterizedTypeToken(nodeKind)) {
+                if (isParameterizedTypeToken(nextTokenKind)) {
                     return true;
                 }
                 
-                if (isSingletonTypeDescStart(nodeKind)) {
+                if (isSingletonTypeDescStart(nextTokenKind, nextNextToken)) {
                     return true;
                 }
-                return isSimpleType(nodeKind);
+                return isSimpleType(nextTokenKind);
         }
     }
 
@@ -10641,7 +10645,7 @@ public class BallerinaParser extends AbstractParser {
         List<STNode> varDecls = new ArrayList<>();
         STToken nextToken = peek();
 
-        if (isEndOfLetVarDeclarations(nextToken.kind)) {
+        if (isEndOfLetVarDeclarations(nextToken.kind, getNextNextToken())) {
             endContext();
             return STNodeFactory.createEmptyNodeList();
         }
@@ -10653,7 +10657,7 @@ public class BallerinaParser extends AbstractParser {
         // Parse the remaining variable declarations
         nextToken = peek();
         STNode leadingComma;
-        while (!isEndOfLetVarDeclarations(nextToken.kind)) {
+        while (!isEndOfLetVarDeclarations(nextToken.kind, getNextNextToken())) {
             leadingComma = parseComma();
             varDecls.add(leadingComma);
             varDec = parseLetVarDecl(isRhsExpr);
@@ -10665,7 +10669,7 @@ public class BallerinaParser extends AbstractParser {
         return STNodeFactory.createNodeList(varDecls);
     }
 
-    private boolean isEndOfLetVarDeclarations(SyntaxKind tokenKind) {
+    static boolean isEndOfLetVarDeclarations(SyntaxKind tokenKind, STToken nextNextToken) {
         switch (tokenKind) {
             case COMMA_TOKEN:
             case AT_TOKEN:
@@ -10673,7 +10677,7 @@ public class BallerinaParser extends AbstractParser {
             case IN_KEYWORD:
                 return true;
             default:
-                return !isTypeStartingToken(tokenKind);
+                return !isTypeStartingToken(tokenKind, nextNextToken);
         }
     }
 
@@ -12357,8 +12361,7 @@ public class BallerinaParser extends AbstractParser {
         return STNodeFactory.createUnaryExpressionNode(operator, literal);
     }
 
-    private boolean isSingletonTypeDescStart(SyntaxKind tokenKind) {
-        STToken nextNextToken = getNextNextToken();
+    static boolean isSingletonTypeDescStart(SyntaxKind tokenKind, STToken nextNextToken) {
         switch (tokenKind) {
             case STRING_LITERAL_TOKEN:
             case DECIMAL_INTEGER_LITERAL_TOKEN:

--- a/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java
+++ b/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java
@@ -9860,7 +9860,7 @@ public class BallerinaParser extends AbstractParser {
         return isTypeStartingToken(nodeKind, getNextNextToken());
     }
     
-    static boolean isTypeStartingToken(SyntaxKind nextTokenKind, STToken nextNextToken) {
+    private static boolean isTypeStartingToken(SyntaxKind nextTokenKind, STToken nextNextToken) {
         switch (nextTokenKind) {
             case IDENTIFIER_TOKEN:
             case SERVICE_KEYWORD:
@@ -12361,7 +12361,7 @@ public class BallerinaParser extends AbstractParser {
         return STNodeFactory.createUnaryExpressionNode(operator, literal);
     }
 
-    static boolean isSingletonTypeDescStart(SyntaxKind tokenKind, STToken nextNextToken) {
+    private static boolean isSingletonTypeDescStart(SyntaxKind tokenKind, STToken nextNextToken) {
         switch (tokenKind) {
             case STRING_LITERAL_TOKEN:
             case DECIMAL_INTEGER_LITERAL_TOKEN:

--- a/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java
+++ b/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java
@@ -3895,7 +3895,7 @@ public class BallerinaParserErrorHandler extends AbstractParserErrorHandler {
                 if (parentCtx == ParserRuleContext.QUERY_EXPRESSION) {
                     STToken nextToken = this.tokenReader.peek(nextLookahead);
                     STToken nextNextToken = this.tokenReader.peek(nextLookahead + 1);
-                    // This is a special case since parser quite parsing let clauses if 
+                    // This is a special case since parser quit parsing let clauses if 
                     // `isEndOfLetVarDeclarations()` is true
                     if (BallerinaParser.isEndOfLetVarDeclarations(nextToken.kind, nextNextToken)) {
                         return ParserRuleContext.LET_CLAUSE_END;

--- a/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java
+++ b/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java
@@ -3594,7 +3594,6 @@ public class BallerinaParserErrorHandler extends AbstractParserErrorHandler {
                 return ParserRuleContext.OPEN_BRACE;
             case RECEIVE_FIELD_NAME:
                 return ParserRuleContext.COLON;
-
             case WAIT_FIELD_NAME:
                 return ParserRuleContext.WAIT_FIELD_NAME_RHS;
             case ALTERNATE_WAIT_EXPR_LIST_END:
@@ -3608,6 +3607,13 @@ public class BallerinaParserErrorHandler extends AbstractParserErrorHandler {
             case DO_CLAUSE:
                 return ParserRuleContext.DO_KEYWORD;
             case LET_CLAUSE_END:
+                parentCtx = getParentContext();
+                // We can reach here without `LET_CLAUSE_LET_VAR_DECL` ctx for empty let var declarations
+                if (parentCtx == ParserRuleContext.LET_CLAUSE_LET_VAR_DECL) {
+                    endContext();
+                    return ParserRuleContext.QUERY_PIPELINE_RHS;
+                }
+                return ParserRuleContext.QUERY_PIPELINE_RHS;
             case ORDER_CLAUSE_END:
             case JOIN_CLAUSE_END:
                 endContext();
@@ -3624,7 +3630,6 @@ public class BallerinaParserErrorHandler extends AbstractParserErrorHandler {
                 return ParserRuleContext.RETRY_KEYWORD;
             case ROLLBACK_STMT:
                 return ParserRuleContext.ROLLBACK_KEYWORD;
-
             case MODULE_ENUM_DECLARATION:
                 return ParserRuleContext.ENUM_KEYWORD;
             case MODULE_ENUM_NAME:
@@ -3888,6 +3893,13 @@ public class BallerinaParserErrorHandler extends AbstractParserErrorHandler {
             case LET_KEYWORD:
                 parentCtx = getParentContext();
                 if (parentCtx == ParserRuleContext.QUERY_EXPRESSION) {
+                    STToken nextToken = this.tokenReader.peek(nextLookahead);
+                    STToken nextNextToken = this.tokenReader.peek(nextLookahead + 1);
+                    // This is a special case since parser quite parsing let clauses if 
+                    // `isEndOfLetVarDeclarations()` is true
+                    if (BallerinaParser.isEndOfLetVarDeclarations(nextToken.kind, nextNextToken)) {
+                        return ParserRuleContext.LET_CLAUSE_END;
+                    }
                     return ParserRuleContext.LET_CLAUSE_LET_VAR_DECL;
                 } else if (parentCtx == ParserRuleContext.LET_CLAUSE_LET_VAR_DECL) {
                     endContext(); // end let-clause-let-var-decl

--- a/compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/misc/ResiliencyTest.java
+++ b/compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/misc/ResiliencyTest.java
@@ -47,4 +47,9 @@ public class ResiliencyTest extends AbstractMiscTest {
     public void testResiliencyForTokenBufferOverflow() {
         testFile("resiliency/resiliency_source_04.bal", "resiliency/resiliency_assert_04.json");
     }
+
+    @Test
+    public void testResiliencyForRecursiveLetClauseRecovery() {
+        testFile("resiliency/resiliency_source_05.bal", "resiliency/resiliency_assert_05.json");
+    }
 }

--- a/compiler/ballerina-parser/src/test/resources/misc/resiliency/resiliency_assert_05.json
+++ b/compiler/ballerina-parser/src/test/resources/misc/resiliency/resiliency_assert_05.json
@@ -1,0 +1,205 @@
+{
+  "kind": "MODULE_PART",
+  "hasDiagnostics": true,
+  "children": [
+    {
+      "kind": "LIST",
+      "children": []
+    },
+    {
+      "kind": "LIST",
+      "hasDiagnostics": true,
+      "children": [
+        {
+          "kind": "MODULE_VAR_DECL",
+          "hasDiagnostics": true,
+          "children": [
+            {
+              "kind": "LIST",
+              "children": []
+            },
+            {
+              "kind": "TYPED_BINDING_PATTERN",
+              "children": [
+                {
+                  "kind": "VAR_TYPE_DESC",
+                  "children": [
+                    {
+                      "kind": "VAR_KEYWORD",
+                      "trailingMinutiae": [
+                        {
+                          "kind": "WHITESPACE_MINUTIAE",
+                          "value": " "
+                        }
+                      ]
+                    }
+                  ]
+                },
+                {
+                  "kind": "CAPTURE_BINDING_PATTERN",
+                  "children": [
+                    {
+                      "kind": "IDENTIFIER_TOKEN",
+                      "value": "x",
+                      "trailingMinutiae": [
+                        {
+                          "kind": "WHITESPACE_MINUTIAE",
+                          "value": " "
+                        }
+                      ]
+                    }
+                  ]
+                }
+              ]
+            },
+            {
+              "kind": "EQUAL_TOKEN",
+              "trailingMinutiae": [
+                {
+                  "kind": "WHITESPACE_MINUTIAE",
+                  "value": " "
+                }
+              ]
+            },
+            {
+              "kind": "SIMPLE_NAME_REFERENCE",
+              "hasDiagnostics": true,
+              "children": [
+                {
+                  "kind": "IDENTIFIER_TOKEN",
+                  "isMissing": true,
+                  "hasDiagnostics": true,
+                  "trailingMinutiae": [
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "hasDiagnostics": true,
+                        "children": [
+                          {
+                            "kind": "FROM_KEYWORD",
+                            "hasDiagnostics": true,
+                            "diagnostics": [
+                              "ERROR_EXPRESSION_EXPECTED_ACTION_FOUND"
+                            ]
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "VAR_KEYWORD"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "IDENTIFIER_TOKEN",
+                            "value": "a"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "IN_KEYWORD"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "IDENTIFIER_TOKEN",
+                            "value": "m"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "EQUAL_TOKEN"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "WHITESPACE_MINUTIAE",
+                      "value": " "
+                    },
+                    {
+                      "kind": "INVALID_NODE_MINUTIAE",
+                      "invalidNode": {
+                        "kind": "INVALID_TOKEN_MINUTIAE_NODE",
+                        "children": [
+                          {
+                            "kind": "OPEN_BRACE_TOKEN"
+                          }
+                        ]
+                      }
+                    },
+                    {
+                      "kind": "END_OF_LINE_MINUTIAE",
+                      "value": "\n"
+                    }
+                  ]
+                }
+              ]
+            },
+            {
+              "kind": "SEMICOLON_TOKEN",
+              "isMissing": true,
+              "hasDiagnostics": true,
+              "diagnostics": [
+                "ERROR_MISSING_SEMICOLON_TOKEN"
+              ]
+            }
+          ]
+        }
+      ]
+    },
+    {
+      "kind": "EOF_TOKEN"
+    }
+  ]
+}

--- a/compiler/ballerina-parser/src/test/resources/misc/resiliency/resiliency_source_05.bal
+++ b/compiler/ballerina-parser/src/test/resources/misc/resiliency/resiliency_source_05.bal
@@ -1,0 +1,1 @@
+var x = from var a in m = {

--- a/misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java
+++ b/misc/formatter/modules/formatter-core/src/test/java/org/ballerinalang/formatter/core/ParserTestFormatter.java
@@ -172,7 +172,7 @@ public class ParserTestFormatter extends FormatterTest {
                 "predeclared-module-prefix_02.bal", "object_type_def_source_44.bal", "record_type_def_source_27.bal",
                 "func_type_source_09.bal", "func_type_source_13.bal", "func_type_source_14.bal",
                 "func_type_source_15.bal", "func_type_source_16.bal", "import_decl_source_24.bal",
-                "member_access_expr_source_11.bal", "float_literal_source_08.bal");
+                "member_access_expr_source_11.bal", "float_literal_source_08.bal", "resiliency_source_05.bal");
     }
 
     @DataProvider(name = "test-file-provider")

